### PR TITLE
[oximeter] Add minimal versioning support

### PR DESCRIPTION
**Pull request by smklein**
Provides a mechanism to explicitly version Oximeter schema used within Clickhouse, and to wipe that data
if any schema-breaking changes are made.

This is retroactively intended to solve the problems discussed within https://github.com/oxidecomputer/omicron/issues/4221 , where the `timeseries_key` generation was altered.

### CODE REVIEW

File: oximeter/db/src/client.rs [hunk 1]
```diff
@@ -1339,6 +1429,126 @@ mod tests {
...
+            )
+            .await
+            .expect("Failed to SELECT from database")
+            .lines()
+            .count()
+    }
+
+    #[tokio::test]
+    async fn test_database_version_update_idempotent() {
+        let log = slog::Logger::root(slog::Discard, o!());
```
Comment (bnaecker) [inline oximeter/db/src/client.rs:1446]: I've been trying to move all the tests over to the test logging infrastructure from `omicron_test_utils`. The original tests predated all that stuff, I think, or I didn't use it for some reason.

Would you mind using the `test_setup_log` function a la L1312?

File: oximeter/db/src/client.rs [hunk 2]
```diff
@@ -1339,6 +1429,126 @@ mod tests {
...
+            .expect("Failed to initialize timeseries database");
+
+        assert_eq!(1, get_schema_count(&client).await);
+
+        db.cleanup().await.expect("Failed to cleanup ClickHouse server");
+    }
+
+    #[tokio::test]
+    async fn test_database_version_will_not_downgrade() {
+        let log = slog::Logger::root(slog::Discard, o!());
```
Comment (bnaecker) [inline oximeter/db/src/client.rs:1483]: Same note about `test_setup_log` here.

File: oximeter/db/src/client.rs [hunk 3]
```diff
@@ -1339,6 +1429,126 @@ mod tests {
...
+            .initialize_db_with_version(replicated, model::OXIMETER_VERSION)
+            .await
+            .expect_err("Should have failed, downgrades are not supported");
+
+        db.cleanup().await.expect("Failed to cleanup ClickHouse server");
+    }
+
+    #[tokio::test]
+    async fn test_database_version_wipes_old_version() {
+        let log = slog::Logger::root(slog::Discard, o!());
```
Comment (bnaecker) [inline oximeter/db/src/client.rs:1518]: Same note about `test_setup_log` here.

**Review by bnaecker — COMMENTED** [review]
Looks good, thanks for the quick and helpful fix on this. Couple of small suggestions. Thanks!

File: oximeter/db/preprocessed_configs/config.xml [hunk 4]
```diff
@@ -0,0 +1,41 @@
+<!-- This file was generated automatically.
```
[the remark this answers is not in the document]
Comment (smklein) [reply oximeter/db/preprocessed_configs/config.xml]: Removed

File: oximeter/db/src/client.rs [hunk 5]
```diff
@@ -269,7 +271,95 @@ impl Client {
         .map_err(|e| Error::Database(e.to_string()))
     }
 
-    // Verifies if instance is part of oximeter_cluster
+    /// Validates that the schema used by the DB matches the version used by
+    /// the executable using it.
+    ///
+    /// This function will wipe metrics data if the version stored within
```
[the remark this answers is not in the document]
Comment (smklein) [reply oximeter/db/src/client.rs:277]: Good catch, updated the comment to cover these cases.

File: oximeter/db/src/client.rs [hunk 6]
```diff
@@ -269,7 +271,95 @@ impl Client {
...
+    }
+
+    async fn read_latest_version(&self) -> Result<u64, Error> {
+        let sql = format!(
+            "SELECT MAX(value) FROM {db_name}.version;",
+            db_name = crate::DATABASE_NAME,
+        );
+
+        let version = match self.execute_with_body(sql).await {
+            Ok(body) if body.is_empty() => {
```
[the remark this answers is not in the document]
Comment (smklein) [reply oximeter/db/src/client.rs:333]: So this case is actually unlikely, but I figured I'd include it regardless.

Here are the cases I'd expect us to use:

- When we upgrade the in-field databases, which exist, but don't have the "version" table, we'll see a database error containing `Table oximeter.version doesn't exist`.
- When we scan for versions in databases that haven't been formatted, we'll see a database error containing `Database oximeter doesn't exist`

This case of "no versions exist" would happen if we formatted the schema, but bailed out before we managed to insert the version (or crashed before we got that far).

I went ahead and updated the code + comments to make this a bit more explicit.

File: oximeter/db/src/client.rs (hunk 1 above)
Comment (smklein) [reply oximeter/db/src/client.rs]: Sounds good, done for this and for several other callers in this file.

**Comment by smklein** [comment]
I just added it in one of the more recent commits, but that check is part of this block:

https://github.com/oxidecomputer/omicron/blob/b3d57654ffdc29cc4f735bb3a6639d0e94bb6d9d/oximeter/db/src/client.rs#L346-L348

**Comment by bnaecker** [comment]
Ah I see it now. Must have been looking at old commits, because of course that’s what GH shows by default. Thanks for addressing everything, feel free to merge when you’re ready.

**Comment by david-crespo** [comment]
This looks good. The thing that makes me nervous is that any Oximeter client (which I understand to be tied to a control plane instance) can basically wipe Clickhouse at any time if it is running with a bumped version? Obviously we would not be bumping this version very often, but I feel there should be more documentation or _something_. At the very least, on the version itself, shouldn't it say something like HI IF YOU CHANGE THIS VALUE YOU WILL DELETE ALL THE DATA.

**Comment by smklein** [comment]
> This looks good. The thing that makes me nervous is that any Oximeter client (which I understand to be tied to a control plane instance) can basically wipe Clickhouse at any time if it is running with a bumped version? Obviously we would not be bumping this version very often, but I feel there should be more documentation or _something_. At the very least, on the version itself, shouldn't it say something like HI IF YOU CHANGE THIS VALUE YOU WILL DELETE ALL THE DATA.

I added some documentation in https://github.com/oxidecomputer/omicron/pull/4246/commits/a6c1798f917a72c5c9ff3b1788d0ca55181f197d , including a link to https://github.com/oxidecomputer/omicron/issues/4271

**Review by david-crespo — APPROVED** [review]

**Pull request merged**